adagrams/game.py

```diff
@@ -1,11 +1,127 @@
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+
+LETTER_SCORE_CHART = {
+    1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+    2: ["D","G"],
+    3: ["B","C","M","P"],
+    4: ["F","H","V","W","Y"],
+    5: ["K"],
+    8: ["J","X"],
+    10:["Q","Z"]
+}
+
+
 def draw_letters():
-    pass
+    # convert the LETTER_POOL into a letter_list which contains all the available letters
+    letter_list = []
+    for (letter, letter_freq) in LETTER_POOL.items():
+        letter_list += [letter] * letter_freq
+    
+    # * pop one random letter from the letter_list and append it to the newly created list(letters), repeat for 10 times.
+    letter_bank =[]
+    for i in range(10):
+        letter = letter_list.pop(random.randrange(len(letter_list)))
+        letter_bank.append(letter)
+    return letter_bank
+
+
+def uses_available_letters(word, letter_bank): 
+    # create two empty dictionaries, in order to convert both the word and letter_bank into dictionaries.
+    letter_bank_dict = {}
+    word_dict ={}
+    # convert letter_bank(list) into a dictionary
+    for letter in letter_bank:
+        if letter in letter_bank_dict:
+            letter_bank_dict[letter] += 1
+        else:
+            letter_bank_dict[letter] = 1
+    
+    # convert word(string) into a dictionary
+    for char in word.upper():
+        if char in word_dict:
+            word_dict[char] += 1
+        else:
+            word_dict[char] = 1
+    
+    # check if every letter in the word is also in the letter_bank dictionary and of right quantity. 
+    for word_letter in word_dict:
+        if word_letter not in letter_bank_dict or word_dict[word_letter] > letter_bank_dict[word_letter]:
+            return False
+    return True
+
 
-def uses_available_letters(word, letter_bank):
-    pass
 
 def score_word(word):
-    pass
+    # if word is empty string, return 0
+    if word == "":
+        return 0
+    
+    # caculate the score according to the LETTER_SCORE_CHART
+    total_score = 0
+    for letter in word.upper():
+        for (score, letters) in LETTER_SCORE_CHART.items():
+            if letter in letters:
+                total_score += score
+    # add 8 more points if the length of word is more than 7
+    if len(word) >= 7:
+        total_score += 8
+        
+    return total_score
+
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    # create a dictionary(scores) in which the key is the word, value is its score. 
+    # find out the highest score of all the scores by using max()
+    scores = {}
+    for word in word_list:
+        scores[word] = score_word(word)
+    highest_score = max(scores.values())
+    
+    # find out all the words whose scores are the same as the highest score, and put them in a list.
+    highest_score_words = []
+    for word, score in scores.items():
+        if score == highest_score:
+            highest_score_words.append(word)
```

Comment:
We're holding onto score info for all the words in `scores`, but we really are only concerned with the highest scoring words. Another approach to getting a list of the highest scored words could be to declare variables for the current high score and a list of the words with that score. We could use a for-loop to update the high score as we loop through the words, and either append to or replace the highest score list as we go:
```python
highest_score = 0
winner_list = []

for word in word_list:
    score = score_word(word)

    if score > highest_score:
        highest_score = score
        winner_list = [word]

    elif score == highest_score:
        winner_list.append(word)
```

Reply:
I have to say this solution is so brilliant!!!!!